### PR TITLE
Startup failures under low memory

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Signed-off-by: Alan Jowett <alan.jowett@microsoft.com>

## Description

Failures that were found after enabling low-memory simulation and starting ebpfcore.sys.

## Testing

Manual testing. Issue #1502 tracks adding this to CI/CD.

## Documentation

No.

Resolves: #1503

### CODE REVIEW

File: scripts/deploy-ebpf.ps1.in [hunk 1]
```diff
@@ -261,7 +261,7 @@ if (! $principal.IsInRole([Security.Principal.WindowsBuiltinRole]::Administrator
    exit 1
 }
 
-Enable-VMIntegrationService -VMName "Windows 10 dev environment" -Name "Guest Service Interface"
+Enable-VMIntegrationService -VMName $vm -Name "Guest Service Interface"
```
Comment (dthaler) [inline scripts/deploy-ebpf.ps1.in:264]: Exclude from this PR.  This is part of https://github.com/microsoft/ebpf-for-windows/pull/1482

**Review by dthaler — DISMISSED** [review]

File: libs/execution_context/ebpf_native.c [hunk 2]
```diff
@@ -193,10 +193,12 @@ ebpf_native_release_reference(_In_opt_ ebpf_native_module_t* module)
                 EBPF_TRACELOG_KEYWORD_NATIVE,
                 "ebpf_native_release_reference: all program references released. Unloading module",
                 module->client_id);
-            module_id = (GUID*)ebpf_allocate(sizeof(GUID));
-            if (module_id == NULL) {
-                result = EBPF_NO_MEMORY;
-                goto Done;
+            // Spin waiting for memory pressure to be released.
```
Comment (Alan-Jowett) [inline libs/execution_context/ebpf_native.c:196]: We need a more elegant solution for these cases.

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_native.c:196]: Filed https://github.com/microsoft/ebpf-for-windows/issues/1511

File: libs/execution_context/ebpf_native.c [hunk 3]
```diff
@@ -1349,19 +1363,25 @@ ebpf_native_unload(_In_ const GUID* module_id)
     // the module memory can be freed immediately after the hash table lock is
     // released. Create a copy of the service name to use later to unload driver.
     service_name_length = (wcslen(module->service_name) + 1) * sizeof(wchar_t);
-    service_name = ebpf_allocate(service_name_length);
-    if (service_name == NULL) {
-        result = EBPF_NO_MEMORY;
-        goto Done;
+
+    // Spin until we can allocate memory for the service name.
+    for (;;) {
```
[the remark this answers is not in the document]
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_native.c:1368]: Filed https://github.com/microsoft/ebpf-for-windows/issues/1511, will back out the spin (it was mostly to get past this failure).

**Review by dthaler — APPROVED** [review]

File: libs/execution_context/ebpf_native.c [hunk 4]
```diff
@@ -208,10 +209,15 @@ ebpf_native_release_reference(_In_opt_ ebpf_native_module_t* module)
             ebpf_free(module_id);
         }
     } else if (new_ref_count == 0) {
+        // TODO: https://github.com/microsoft/ebpf-for-windows/issues/1511
```
Comment (saxena-anurag) [inline libs/execution_context/ebpf_native.c:212]: I think this should be issue #1506 ?

**Review by saxena-anurag — DISMISSED** [review]

**Review by dthaler — APPROVED** [review]

File: libs/platform/ebpf_pinning_table.c [hunk 5]
```diff
@@ -204,17 +207,22 @@ ebpf_pinning_table_delete(ebpf_pinning_table_t* pinning_table, const ebpf_utf8_s
     if (return_value == EBPF_SUCCESS) {
         entry = *existing_pinning_entry;
         return_value = ebpf_hash_table_delete(pinning_table->hash_table, (const uint8_t*)&existing_key);
+        // If unable to remove the entry from the table, don't delete it.
+        if (return_value != EBPF_SUCCESS) {
```
Comment (shankarseal) [inline libs/platform/ebpf_pinning_table.c:211]: How can hash_table_delete fail unless the key is not found? And you are holding the pinning_table lock here.

Comment (Alan-Jowett) [reply libs/platform/ebpf_pinning_table.c:211]: The hash table implements RCU semantics. This means that on the deletion of a key it allocates a new block to hold the pointers to hashtable buckets. The allocation of the new block can fail, which causes the table update to fail.

**Review by saxena-anurag — APPROVED** [review]

**Pull request merged**